src/util/validateTopic.ts

```diff
@@ -1,39 +1,61 @@
-import { logger } from './logger.js';
-
-export function validateTopic(topic: string, message: string): boolean {
-  const end = topic.length - 1;
-  const endMinus = end - 1;
-  const slashInPreEnd = endMinus > 0 && topic.charCodeAt(endMinus) !== 47;
+/**
+ * Validate a topic to see if it's valid or not.
+ * Topics can support Multi-level wildcards (‘#’ U+0023) and single-level wildcards (‘+’ U+002B).
+ * 
+ * The multi-level wildcard character MUST be specified either on its own or following a topic level separator.
+ * In either case it MUST be the last character specified in the Topic Filter.
+ * 
+ * The single-level wildcard character MUST occupy an entire level of the filter.
+ * For example: "+" is valid, "+/tennis/#" is valid, "sport+" is not valid.
+ * 
+ * Topics beginning with $: The Server MUST NOT match Topic Filters starting with a wildcard character with Topic Names
+ * 
+ * Topic names and topic filters must not include the null character (Unicode U+0000)
+ * 
+ * Topic names and Topic Filters are UTF-8 encoded strings; they must not encode to more than 65,535 bytes.
+ *
+ * @param {String} topic - A topic
+ * @returns {Boolean} If the topic is valid, returns true. Otherwise, returns false.
+ */
+export function validateTopic(topic: string) {
+  // Topic must be at least 1 character.
   if (topic.length === 0) {
-    // [MQTT-3.8.3-3]
-    const err = new Error('impossible to ' + message + ' to an empty topic');
-    logger.info(err);
     return false;
   }
-  for (let i = 0; i < topic.length; i++) {
-    switch (topic.charCodeAt(i)) {
-      case 35: {
-        // #
-        const notAtTheEnd = i !== end;
-        if (notAtTheEnd || slashInPreEnd) {
-          const err = new Error('# is only allowed in ' + message + ' in the last position');
-          logger.info(err);
-          return false;
-        }
-        break;
-      }
-      case 43: {
-        // +
-        const pastChar = i < end - 1 && topic.charCodeAt(i + 1) !== 47;
-        const preChar = i > 1 && topic.charCodeAt(i - 1) !== 47;
-        if (pastChar || preChar) {
-          const err = new Error('+ is only allowed in ' + message + ' between /');
-          logger.info(err);
-          return false;
-        }
-        break;
-      }
+  const levels: string[] = topic.split('/');
+
+  for (let i = 0; i < levels.length; i++) {
+    // If SLWC, MUST occupy entire level.
+    if (levels[i] === '+') {
+      continue;
+    }
+
+    if (levels[i] === '#') {
+      // Validate MLWC at end of topic filter.
+      return i === levels.length - 1;
+    }
+
+    // Level must not contain more than one MLWC or SLWC character.
+    if ((levels[i] as string).indexOf('+') !== -1 || (levels[i] as string).indexOf('#') !== -1) {
+      return false;
     }
   }
```

Comment:
Haven't verified if this builds but this makes things a bit more readable IMO
```suggestion
  for (const [i, level] of levels.entries()) {
    // If SLWC, MUST occupy entire level.
    if (level === '+') {
      continue;
    }

    if (level === '#') {
      // Validate MLWC at end of topic filter.
      return i === levels.length - 1;
    }

    // Level must not contain more than one MLWC or SLWC character.
    if (level.includes('+') || levels.includes('#')) {
      return false;
    }
  }
```